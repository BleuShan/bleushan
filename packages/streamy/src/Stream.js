import {invariant} from './utils/invariant.js'
import {isArrayLike} from './utils/isArrayLike.js'
import {isString} from './utils/isString.js'
/* eslint-disable no-unused-vars */
import {typeOf} from './utils/typeOf.js'
import {IndexIterator} from './iterators/IndexIterator.js'
/* eslint-enable no-unused-vars */

export class Stream {
  #completed = false
  #isAsync
  #memoizedResults = []
  #processResult = ({value, done}) => {
    this.#completed = done && value == null

    if (!this.#completed) {
      this.#memoizedResults.push(value)
    } else {
      this.#sourceIterator = null
    }

    return {
      value,
      done: this.#completed
    }
  }

  #source
  #sourceIterator

  // eslint-disable-next-line no-undef
  get #iterator() {
    if (this.#sourceIterator == null) {
      if (this.#completed) {
        this.#sourceIterator = new IndexIterator(this.#memoizedResults)
      } else if (this.isAsync) {
        this.#sourceIterator = this.#source[Symbol.asyncIterator]()
      } else if (typeOf(this.#source[Symbol.iterator]) === 'function') {
        this.#sourceIterator = this.#source[Symbol.iterator]()
      } else {
        this.#sourceIterator = new IndexIterator(this.#source)
      }
    }

    return this.#sourceIterator
  }

  get isAsync() {
    return this.#isAsync
  }

  constructor(source) {
    invariant({
      condition: isArrayLike(source) || isString(source),
      message: 'source must be a string, an ArrayLike object or an iterable',
      errorType: TypeError
    })

    this.#source = source
    this.#isAsync = typeOf(this.#source[Symbol.asyncIterator]) === 'function'
  }

  [Symbol.asyncIterator]() {
    return this
  }

  [Symbol.iterator]() {
    invariant({
      condition: !this.isAsync || this.#completed,
      message: 'Cannot iterate synchronously over an uncompleted asynchronous stream',
      errorType: TypeError
    })

    return this
  }

  next(input) {
    const promiseOrResult = this.#iterator.next(input)
    if (!this.#completed) {
      if (typeOf(promiseOrResult.then) === 'function') {
        return promiseOrResult.then(this.#processResult)
      } else {
        return this.#processResult(promiseOrResult)
      }
    }

    return promiseOrResult
  }
}
