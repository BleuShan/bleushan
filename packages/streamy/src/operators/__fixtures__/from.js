export function* generator() {
  yield 1
  yield 2
  yield 3
}

export async function* asyncGenerator() {
  for (let i = 1; i < 10; i++) {
    const result = await new Promise((resolve) => {
      resolve(i * 2)
    })
    yield result
  }
}

export class AsyncIterable {
  #count
  constructor() {
    this.#count = 0
  }

  [Symbol.asyncIterator]() {
    return this
  }

  async next() {
    const result = {
      done: this.#count >= 10
    }

    if (!result.done) {
      result.value = ++this.#count
    }

    return result
  }

  reset() {
    this.#count = 0
  }
}

export class EchoIterable {
  [Symbol.iterator]() {
    return this
  }

  next(...value) {
    return !value.length
      ? {
          done: true
        }
      : {
          value,
          done: false
        }
  }
}
