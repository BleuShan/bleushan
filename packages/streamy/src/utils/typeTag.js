import {toStringTag} from './internal/toStringTag.js'

const extractor = /\[object (\w+)\]/

export function typeTag(target) {
  return toStringTag(target).replace(extractor, '$1')
}
