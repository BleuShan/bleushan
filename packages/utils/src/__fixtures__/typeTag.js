export const tagged = Object.create(null, {
  name: {
    value: 'tagged',
    enumerable: true
  },
  [Symbol.toStringTag]: {
    value: 'StringTagged'
  }
})

export const inheritedTag = Object.create(tagged, {
  name: {
    value: 'inheritedTag',
    enumerable: true
  }
})

export const writableTagged = Object.create(null, {
  name: {
    value: 'writableTagged',
    enumerable: true
  },
  [Symbol.toStringTag]: {
    value: 'StringTagged',
    writable: true
  }
})

export const inheritedWritableTag = Object.create(writableTagged, {
  name: {
    value: 'inheritedWritableTag',
    enumerable: true
  }
})
