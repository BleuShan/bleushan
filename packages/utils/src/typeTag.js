import toStringTag from './internal/toStringTag.js'

const extractor = /\[object (\w+)\]/

export default function typeTag(target) {
  return toStringTag(target).replace(extractor, '$1')
}
