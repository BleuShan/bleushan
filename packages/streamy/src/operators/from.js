import {Stream} from '../Stream.js'

export function from(source) {
  return new Stream(source)
}
