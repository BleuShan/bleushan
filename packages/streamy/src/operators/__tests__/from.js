import {from} from '../from.js'
import {Stream} from '../../Stream.js'
import {asyncGenerator, AsyncIterable, generator} from '../__fixtures__/from.js'
import {typeTag} from '../../utils/typeTag.js'

describe('from operator', () => {
  describe.each([
    [[1, 2, 3], true],
    [{0: 1, length: 1}, true],
    ['sast', true],
    [new Set(['s']), true],
    [generator, true],
    [asyncGenerator, true],
    [new AsyncIterable(), true],
    [({test: 's'}, false)],
    [null, false],
    [undefined, false],
    [NaN, false]
  ])('when called on %p', (source, valid) => {
    let call
    let expectedResult

    beforeEach(async () => {
      const isGenerator =
        typeTag(source) === 'GeneratorFunction' || typeTag(source) === 'AsyncGeneratorFunction'
      if (valid) {
        const iter = isGenerator ? source() : source
        if (iter[Symbol.asyncIterator]) {
          expectedResult = []
          for await (const value of iter) {
            expectedResult.push(value)
          }
        } else {
          expectedResult = Array.from(iter)
        }
      }

      call = () => (isGenerator ? from(source()) : from(source))
    })

    if (valid) {
      it('should not throw', () => {
        expect(call).not.toThrow()
      })

      it('should return a Stream', () => {
        expect(call()).toBeInstanceOf(Stream)
      })

      it('should yield the same values as the source', async () => {
        const result = []
        const stream = call()
        if (stream.isAsync) {
          for await (const value of stream) {
            result.push(value)
          }
        } else {
          for (const value of stream) {
            result.push(value)
          }
        }

        expect(result).toEqual(expectedResult)
      })
    } else {
      it('should throw', () => {
        expect(call).toThrowErrorMatchingSnapshot()
      })
    }
  })

  describe('when it returns an async stream', () => {
    let stream
    const iter = new AsyncIterable()
    let result

    const iterate = () => {
      for (const value of stream) {
        if (result == null) {
          result = []
        }
        result.push(value)
      }
    }

    beforeEach(() => {
      stream = from(iter)
    })

    afterEach(() => {
      result = null
      jest.restoreAllMocks()
      iter.reset()
    })

    it('should not allow synchronous calls if has not completed yet', () => {
      expect(iterate).toThrowErrorMatchingSnapshot()
      expect(result).toBeNil()
    })

    it('should allow synchronous calls if has completed', async () => {
      const expected = []
      for await (const value of stream) {
        expected.push(value)
      }

      jest.spyOn(iter, 'next')

      expect(iterate).not.toThrow()
      expect(iter.next).not.toHaveBeenCalled()
      expect(result).toEqual(expected)
    })
  })
})
