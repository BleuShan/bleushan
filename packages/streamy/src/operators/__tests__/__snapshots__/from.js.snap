// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`from operator when called on NaN should throw 1`] = `"source must be a string, an ArrayLike object or an iterable"`;

exports[`from operator when called on false should throw 1`] = `"source must be a string, an ArrayLike object or an iterable"`;

exports[`from operator when called on null should throw 1`] = `"source must be a string, an ArrayLike object or an iterable"`;

exports[`from operator when called on undefined should throw 1`] = `"source must be a string, an ArrayLike object or an iterable"`;

exports[`from operator when it returns an async stream should not allow synchronous calls if has not completed yet 1`] = `"Cannot iterate synchronously over an uncompleted asynchronous stream"`;
